src/main/java/org/opensearch/knn/index/codec/nativeindex/MemOptimizedNativeIndexBuildStrategy.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.nativeindex;
+
+import org.opensearch.knn.index.VectorDataType;
+import org.opensearch.knn.index.codec.nativeindex.model.BuildIndexParams;
+import org.opensearch.knn.index.codec.transfer.OffHeapByteVectorTransfer;
+import org.opensearch.knn.index.codec.transfer.OffHeapFloatVectorTransfer;
+import org.opensearch.knn.index.codec.transfer.VectorTransfer;
+import org.opensearch.knn.index.engine.KNNEngine;
+import org.opensearch.knn.index.vectorvalues.KNNFloatVectorValues;
+import org.opensearch.knn.index.vectorvalues.KNNVectorValues;
+import org.opensearch.knn.jni.JNIService;
+
+import java.io.IOException;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Map;
+
+/**
+ * Iteratively builds the index.
+ */
+final class MemOptimizedNativeIndexBuildStrategy implements NativeIndexBuildStrategy {
+
+    private static MemOptimizedNativeIndexBuildStrategy INSTANCE = new MemOptimizedNativeIndexBuildStrategy();
+
+    public static MemOptimizedNativeIndexBuildStrategy getInstance() {
+        return INSTANCE;
+    }
+
+    private MemOptimizedNativeIndexBuildStrategy() {}
+
+    public boolean buildIndex(BuildIndexParams indexInfo, final KNNVectorValues<?> knnVectorValues) throws IOException {
+        if (knnVectorValues.docId() == -1) {
```

Comment:
Its being defensive if this class is used standalone

Reply:
its okay to be defensive, but we will loose this context in sometime. So anytime we are adding a defensive check its better to add the java docs to tell that this is defensive check and where it is already handled. Otherwise what will happen is we will just forget why we did this.

Another thing is if we can make this function only accessible internally we can avoid these checks.

Reply:
will add javadocs, I encapsulated this in KNNVectorValues. let me know what you think about it